lightning/src/ln/outbound_payment.rs

```diff
@@ -551,145 +563,208 @@ impl OutboundPayments {
+		if let Err(e) = res {
+			self.handle_pay_route_err(e, payment_id, payment_hash, route, route_params, router, first_hops, inflight_htlcs, entropy_source, node_signer, best_block_height, logger, pending_events, send_payment_along_path);
+		}
+	}
+
+	fn handle_pay_route_err<R: Deref, NS: Deref, ES: Deref, IH, SP, L: Deref>(
+		&self, err: PaymentSendFailure, payment_id: PaymentId, payment_hash: PaymentHash, route: Route,
+		mut route_params: RouteParameters, router: &R, first_hops: Vec<ChannelDetails>,
+		inflight_htlcs: &IH, entropy_source: &ES, node_signer: &NS, best_block_height: u32, logger: &L,
+		pending_events: &Mutex<Vec<events::Event>>, send_payment_along_path: &SP,
+	)
+	where
+		R::Target: Router,
+		ES::Target: EntropySource,
+		NS::Target: NodeSigner,
+		L::Target: Logger,
+		IH: Fn() -> InFlightHtlcs,
+		SP: Fn(&Vec<RouteHop>, &Option<PaymentParameters>, &PaymentHash, &Option<PaymentSecret>, u64,
+		    u32, PaymentId, &Option<PaymentPreimage>, [u8; 32]) -> Result<(), APIError>
+	{
+		match err {
+			PaymentSendFailure::AllFailedResendSafe(errs) => {
+				push_path_failed_evs_and_scids(payment_id, payment_hash, Some(&mut route_params), route.paths, errs.into_iter().map(|e| Err(e)), pending_events);
+				self.retry_payment_internal(payment_id, route_params, router, first_hops, inflight_htlcs, entropy_source, node_signer, best_block_height, logger, pending_events, send_payment_along_path);
+			},
+			PaymentSendFailure::PartialFailure { failed_paths_retry: Some(retry), results, .. } => {
+				push_path_failed_evs_and_scids(payment_id, payment_hash, None, route.paths, results.into_iter(), pending_events);
+				// Some paths were sent, even if we failed to send the full MPP value our recipient may
+				// misbehave and claim the funds, at which point we have to consider the payment sent, so
+				// return `Ok()` here, ignoring any retry errors.
+				self.retry_payment_internal(payment_id, retry, router, first_hops, inflight_htlcs, entropy_source, node_signer, best_block_height, logger, pending_events, send_payment_along_path);
+			},
+			PaymentSendFailure::PartialFailure { failed_paths_retry: None, .. } => {
+				// This may happen if we send a payment and some paths fail, but only due to a temporary
+				// monitor failure or the like, implying they're really in-flight, but we haven't sent the
+				// initial HTLC-Add messages yet.
+			},
+			PaymentSendFailure::PathParameterError(results) => {
+				push_path_failed_evs_and_scids(payment_id, payment_hash, None, route.paths, results.into_iter(), pending_events);
+				self.abandon_payment(payment_id, pending_events);
+			},
+			PaymentSendFailure::ParameterError(e) => {
+				log_error!(logger, "Failed to send to route due to parameter error: {:?}. Your router is buggy", e);
+				self.abandon_payment(payment_id, pending_events);
+			},
+			PaymentSendFailure::DuplicatePayment => {}, // unreachable
```

Comment:
Perhaps add `unreachable!()` as a guard to prevent future refactoring silently introducing a bug
```suggestion
			PaymentSendFailure::DuplicatePayment => { unreachable!() },
```

Reply:
Ah, that commit is from the PR that this PR is based on, #2014. It's fixed there.